### DISCUSSION TITLE
WARNING: Addrspace is not set

### DISCUSSION THREAD

**Discussion by yangxingwu** [bpftrace/bpftrace#2422] [Q&A] [answered]
```bash
$ sudo bpftrace -e 'BEGIN { printf("%x\n", kaddr("usbcore_name")); exit(); }'
Attaching 1 probe...
a7164138

$ sudo bpftrace -e 'BEGIN { printf("%x\n", *kaddr("usbcore_name")); exit(); }'
Attaching 1 probe...
a6699f4e

$ sudo bpftrace -e 'BEGIN { printf("%s\n", str(*kaddr("usbcore_name"))); exit(); }'
WARNING: Addrspace is not set
Attaching 1 probe...
usbcore
```

but if I run `sudo bpftrace -e 'BEGIN { printf("%s\n", str((uint64 *)0xa6699f4e)); exit(); }'`, it prints nothing and **WARNING: Addrspace is not set**. WHY?

```bash
$ sudo bpftrace -e 'BEGIN { printf("%s\n", str((uint64 *)0xa6699f4e)); exit(); }'
WARNING: Addrspace is not set
Attaching 1 probe...
```

I got two questions:

1. what does `kaddr()` actually return?
2. why did not `BEGIN { printf("%s\n", str((uint64 *)0xa6699f4e)); exit(); }` works? and how to fix it?

**Comment by fbs** [comment, marked as answer]
You can find the call signtures in the man page. Kaddr is defined as `uint64 kaddr(const string name)` with the address space set to `kernel`. Which itself is a bit weird as an uint64 shouldn't have an addrspace but it's a left over from a weaker type system. Because you're dereferencing the value (`*kaddr(...)`) that address space info is lost

> why did not `'BEGIN { printf("%s\n", str((uint64 *)0xa6699f4e)); exit(); }' works? and how to fix it?

why a `uint64` pointer and not an `uint8`?

**Reply by yangxingwu** [reply]
thanks @fbs 

> why a `uint64` pointer and not an `uint8`?

An `uint8` pointer also NOT works.

**Reply by yangxingwu** [reply]
thanks @fbs 

use `kptr()` to annotate a pointer belonging to kernel address space works as expected.

```bash
$ sudo bpftrace -e 'BEGIN { $x = *kaddr("usbcore_name"); kptr($x); printf("%s\n", str($x)); exit(); }'
WARNING: Addrspace is not set
Attaching 1 probe...
usbcore
```
